libkbfs/folder_branch_ops.go

```diff
@@ -593,15 +660,82 @@ func (fbo *folderBranchOps) setHeadLocked(ctx context.Context,
 			fbo.identifyDone = false
 		}()
 	}
-	if !wasReadable && md.IsReadable() {
-		// Let any listeners know that this folder is now readable,
-		// which may indicate that a rekey successfully took place.
-		fbo.config.Reporter().Notify(ctx, mdReadSuccessNotification(
-			md.GetTlfHandle().GetCanonicalName(), md.ID.IsPublic()))
+
+	return nil
+}
+
+// setHeadPredecessorLocked is for when we're unstaging updates.
+func (fbo *folderBranchOps) setHeadPredecessorLocked(ctx context.Context,
+	lState *lockState, md *RootMetadata) error {
+	fbo.mdWriterLock.AssertLocked(lState)
+	fbo.headLock.AssertLocked(lState)
+	if fbo.head == nil {
+		return errors.New("Unexpected nil head in setHeadPredecessorLocked")
+	}
+	if fbo.head.Revision <= MetadataRevisionInitial {
+		return fmt.Errorf("setHeadPredecessorLocked unexpectedly called with revision %d", fbo.head.Revision)
+	}
+	err := md.CheckValidSuccessor(fbo.config, fbo.head)
+	if err != nil {
+		return err
+	}
+
+	oldHandle := fbo.head.GetTlfHandle()
+	newHandle := md.GetTlfHandle()
+
+	// The new, older handle might become less resolved.
+	resolvesTo, partialResolvedNewHandle, err :=
+		newHandle.ResolvesTo(
+			ctx, fbo.config.Codec(), fbo.config.KBPKI(),
+			oldHandle)
+	if err != nil {
+		return err
+	}
+
+	oldName := oldHandle.GetCanonicalName()
+	newName := newHandle.GetCanonicalName()
+
+	if !resolvesTo {
+		return IncompatibleHandleError{
+			oldName,
+			partialResolvedNewHandle.GetCanonicalName(),
+			newName,
+		}
 	}
+
+	err = fbo.setHeadLocked(ctx, lState, md)
+	if err != nil {
+		return err
+	}
+
+	if oldName != newName {
```

Comment:
Ah, cool, that simplifies things. Added a check for unmerged head, and changed the check to check for equality.